test/runtests.jl

```diff
@@ -263,23 +265,151 @@ end
+        S3.put(bucket, "test.csv", codeunits(multicsv); credentials)
+        obj = CloudStore.Object(bucket, "test.csv"; credentials)
+        @test length(obj) == sizeof(multicsv)
+
+        N = 1024*1024
+        buf = Vector{UInt8}(undef, N)
+        copyto!(buf, 1, obj, 1, N)
+        @test buf == view(codeunits(multicsv), 1:N)
+
+        ioobj = CloudStore.PrefechedDownloadStream(bucket, "test.csv", 1024*1024; credentials)
+        i = 1
+        while i < sizeof(multicsv)
+            nb = i + N > length(multicsv) ? length(multicsv) - i : N
+            unsafe_read(ioobj, pointer(buf), nb)
+            @test view(buf, 1:nb) == view(codeunits(multicsv), i:i+nb-1)
+            i += N
+        end
+    end
+end
+
+
+
+@testset "CloudStore.PrefechedDownloadStream small readbytes! decompress" begin
+    Minio.with(; debug=true) do conf
+        credentials, bucket = conf
+        multicsv = "1,2,3,4,5,6,7,8,9,1\n"^100; # 2000 B
+        codec = ZlibCompressor();
+        CodecZlib.initialize(codec)
+        compressed = transcode(codec, codeunits(multicsv))
+        S3.put(bucket, "test.csv.gz", compressed; credentials)
+        CodecZlib.finalize(codec)
+        obj = CloudStore.Object(bucket, "test.csv.gz"; credentials)
+        @test length(obj) == sizeof(compressed)
+
+        N = 19
+        buf = Vector{UInt8}(undef, N)
+        ioobj = GzipDecompressorStream(CloudStore.PrefechedDownloadStream(bucket, "test.csv.gz", 16; credentials))
+        i = 1
+        while i < sizeof(multicsv)
+            nb = i + N > length(multicsv) ? length(multicsv) - i : N
+            readbytes!(ioobj, buf, N)
+            @test view(buf, 1:nb) == view(codeunits(multicsv), i:i+nb-1)
+            i += N
+        end
+    end
+end
+
+@testset "CloudStore.PrefechedDownloadStream large readbytes! decompress" begin
+    Minio.with(; debug=true) do conf
+        credentials, bucket = conf
+        multicsv = "1,2,3,4,5,6,7,8,9,1\n"^1000000; # 20 MB
+        codec = ZlibCompressor();
+        CodecZlib.initialize(codec)
+        compressed = transcode(codec, codeunits(multicsv))
+        S3.put(bucket, "test.csv.gz", compressed; credentials)
+        CodecZlib.finalize(codec)
+        obj = CloudStore.Object(bucket, "test.csv.gz"; credentials)
+        @test length(obj) == sizeof(compressed)
+
+        N = 1024*1024
+        buf = Vector{UInt8}(undef, N)
+        ioobj = GzipDecompressorStream(CloudStore.PrefechedDownloadStream(bucket, "test.csv.gz", 16*1024; credentials))
         i = 1
-        while !eof(ioobj)
-            readbytes!(ioobj, buf, 1000)
-            @test buf == view(codeunits(multicsv), i:min(i+999, length(multicsv)))
-            i += 1000
+        while i < sizeof(multicsv)
+            nb = i + N > length(multicsv) ? length(multicsv) - i : N
+            readbytes!(ioobj, buf, N)
+            @test view(buf, 1:nb) == view(codeunits(multicsv), i:i+nb-1)
+            i += N
         end
     end
 end
+end # @testset "CloudStore.jl"
```

Comment:
Can we please add some tests that thoroughly test all the corner cases of the concurrency? I think directly testing what will happen if any of the HTTP calls throw an exception, and testing some corner cases of the possible ordering of the responses.

Probably the best way to do that would be with Mocking. You can see some examples that do this decently well in the raicode protocol tests, in case that's helpful: https://github.com/RelationalAI/raicode/blob/1660f17e12724686f859c2b326ac5e79dfca6773/test/Server/v2-api.jl#L99